usr/share/rear/lib/rsync-functions.sh

```diff
@@ -0,0 +1,163 @@
+# Functions for manipulation of rsync URLs (both OUTPUT_URL and BACKUP_URL)
+
+#### OLD STYLE:
+# BACKUP_URL=[USER@]HOST:PATH           # using ssh (no rsh)
+#
+# with rsync protocol PATH is a MODULE name defined in remote /etc/rsyncd.conf file
+# BACKUP_URL=[USER@]HOST::PATH          # using rsync
+# BACKUP_URL=rsync://[USER@]HOST[:PORT]/PATH    # using rsync (is not compatible with new style!!!)
+
+#### NEW STYLE:
+# BACKUP_URL=rsync://[USER@]HOST[:PORT]/PATH    # using ssh
+# BACKUP_URL=rsync://[USER@]HOST[:PORT]::/PATH  # using rsync
+
+function rsync_validate () {
+    local url="$1"
+
+    if [[ "$(url_scheme "$url")" != "rsync" ]]; then # url_scheme still recognizes old style
+        Error "Non-rsync URL $url !"
+    fi
+}
+
+# Determine whether the URL specifies the use of the rsync protocol (rsyncd) or ssh
+function rsync_proto () {
+    local url="$1"
+
+    rsync_validate "$url"
+    if egrep -q '(::)' <<< $url ; then # new style '::' means rsync protocol
+        echo rsync
+    else
+        echo ssh
+    fi
+}
+
+# Functions to parse the URL into its components:
+# USER, HOST, PORT, PATH
+
+function rsync_user () {
+    local url="$1"
+    local host
+
+    host=$(url_host "$url")
+
+    if grep -q '@' <<< $host ; then
+        echo "${host%%@*}"    # grab user name
+    else
+        echo root
+    fi
+}
+
+function rsync_host () {
+    local url="$1"
+    local host
+    local path
+
+    host=$(url_host "$url")
+    path=$(url_path "$url")
+    # remove USER@ if present
+    local tmp2="${host#*@}"
+
+    case "$(rsync_proto "$url")" in
+        (rsync)
+            # tmp2=witsbebelnx02::backup or tmp2=witsbebelnx02::
+            echo "${tmp2%%::*}"
+            ;;
+        (ssh)
+            # tmp2=host or tmp2=host:
+            echo "${tmp2%%:*}"
+            ;;
+    esac
+}
+
+function rsync_path () {
+    local url="$1"
+    local host
+    local path
+
+    host=$(url_host "$url")
+    path=$(url_path "$url")
+    local tmp2="${host#*@}"
+
+    case "$(rsync_proto "$url")" in
+
+        (rsync)
+            # path=/gdhaese1@witsbebelnx02::backup or path=/backup
+            if grep -q '::' <<< $path ; then
+                echo "${path##*::}"
+            else
+                # XXX what if path=/backup/sub/directory ? Should we remove
+                # longest prefix?
+                echo "${path##*/}"
+            fi
+            ;;
+        (ssh)
+            echo "$path"
+            ;;
+
+    esac
+}
+
+function rsync_port () {
+    # XXX changing port not implemented yet
+    echo 873
```

Comment:
Caution!
Regardless that RSYNC_PORT is not documented in default.conf
it might be used by users who inspected the code and found it
in usr/share/rear/prep/RSYNC/default/100_check_rsync.sh
so I suggest to use something like
```
function rsync_port () {
    # Regardless that RSYNC_PORT is not documented in default.conf
    # it might be used by users who inspected the code up to ReaR 2.6
    # where RSYNC_PORT was used in prep/RSYNC/default/100_check_rsync.sh
    # (873 is the default port of rsync server):
    is_positive_integer $RSYNC_PORT && echo $RSYNC_PORT || echo 873
}
```
or - perhaps even better - we document RSYNC_PORT
officially in default.conf ?

Reply:
This line comes from here: https://github.com/rear/rear/pull/2831/files#diff-e625ae03f11218c80d8b5a4b0f00017895f41de9c6139a12783cd60142df31c7L32 . Note that in the old version of the code, the number was also hardcoded and was not a user-settable parameter. So, the behavior has not changed and it should not become a parameter now. I would prefer to leave a proper implementation of PORT to another PR.